assets/src/edit-story/components/dropzone/index.js

```diff
@@ -0,0 +1,102 @@
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+
+/**
+ * WordPress dependencies
+ */
+import { useRef, useLayoutEffect } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useDropZone from './useDropZone';
+
+const DropZoneComponent = styled.div`
+	display: inherit;
+	position: relative;
+	${ ( { borderPosition, theme, highlightWidth } ) => borderPosition && `
+		:after {
+			height: 100%;
+			display: block;
+			position: absolute;
+			width: ${ highlightWidth }px;
+			background: ${ theme.colors.action };
+			content: '';
+			${ borderPosition }: -${ highlightWidth / 2 }px;
+		}
+	` }
+`;
+
+function DropZone( { children, onDrop } ) {
+	const dropZoneElement = useRef( null );
+	const { actions: { addDropZone, resetHoverState }, state: { hoveredDropZone, dropZones } } = useDropZone();
+
+	useLayoutEffect( () => {
+		if ( ! dropZones.some( ( { ref } ) => ref === dropZoneElement.current ) ) {
+			addDropZone( {
+				ref: dropZoneElement.current,
+			} );
+		}
+	}, [ addDropZone, dropZones ] );
```

Comment:
This is a bit recursive. The side-effect modifies its own dependency. Maybe it'd be easier to just rename `addDropZone` to `registerDropZone` and leave it to its internal implementation? As a benefit, you would also be able to do do something like this:

```
useLayoutAffect( () =>
   registerDropZone(dropZoneElement.current), [registerDropZone] )
```

And the `registerDropZone` could return the unregister function which would allow us to unregister drop zones when a component is unmounted automatically. E.g.

```
function registerDropZone(node) {
  ...
  return () => {
    // unregister here.
  };
}
```

Reply:
Then you should make it a "real" custom hook, call it `useRegisterDropZone` and use it like that. The above is a bit fragile if someone doesn't see that the implicit return of the arrow function is actually important and changes it to:

```js
useLayoutAffect( () => {
  // something here
  registerDropZone(dropZoneElement.current);
}, [registerDropZone] )
```

Now the unregistering doesn't work, because the return value of `registerDropZone` is returned forward.

With a custom hook, it's a lot more clear how it works.

Reply:
Added `unregisterDropZone` and moved the checks to the callbacks directly.